src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsWriteStream.cs

```diff
@@ -0,0 +1,274 @@
+            WriteBufferOffset = TdsEnums.HEADER_LEN;
+        }
+
+        /// <inheritdoc />
+        public override int Read(byte[] buffer, int offset, int count) => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override long Seek(long offset, SeekOrigin origin) => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override void SetLength(long value) => throw new NotSupportedException();
+
+        /// <summary>
+        /// Write operation with the byte buffer is not implemented yet.
+        /// </summary>
+        /// <param name="buffer"></param>
+        /// <param name="offset"></param>
+        /// <param name="count"></param>
+        /// <exception cref="NotSupportedException"></exception>
+        public override void Write(byte[] buffer, int offset, int count) 
+            => Write(buffer.AsSpan(offset, count));
+
+        /// <summary>
+        /// An overload for writing a byte to the stream.
+        /// Use this in case the caller is always going to write asynchronously. 
+        /// Otherwise use the WriteByteAsync because it takes the isAsync flag.
+        /// </summary>
+        /// <param name="value">The byte value to be written to the stream.</param>
+        public override void WriteByte(byte value) => WriteByteAsync(value, isAsync: false, ct: CancellationToken.None).AsTask().Wait();
+
+        /// <summary>
+        /// An async mechanism to write a byte to the stream.
+        /// </summary>
+        /// <param name="value">The byte value to be written</param>
+        /// <param name="isAsync">Indicates if the operation is Async</param>
+        /// <param name="ct">Cancellation token to cancel the async operation.</param>
+        /// <returns></returns>
+        public virtual async ValueTask WriteByteAsync(byte value, bool isAsync, CancellationToken ct)
+        {
+            // If we are already at the end of the buffer, flush the buffer, with a softflush.
+            if (_WriteBuffer.Length - WriteBufferOffset == 0)
+            {
+                await FlushAsync(ct, isAsync, FlushMode.SoftFlush).ConfigureAwait(false);
+            }
+            _WriteBuffer[WriteBufferOffset++] = value;
+        }
+
+        /// <inheritdoc/>
+        public override void Write(ReadOnlySpan<byte> buffer)
+        {
+            int len = buffer.Length;
+            // The buffer may not have enough space. Write what we can and then flush the buffer with a soft flush.
+            while (len > 0)
+            {
+                if (len > _WriteBuffer.Length - WriteBufferOffset)
+                {
+                    // Only a part of the length fits in the buffer.
+                    int bytesToWrite = _WriteBuffer.Length - WriteBufferOffset;
+                    buffer.Slice(0, bytesToWrite).CopyTo(_WriteBuffer.AsSpan(WriteBufferOffset));
+                    WriteBufferOffset += bytesToWrite;
```

Comment:
If this isn't called within a synchronised context, this'll need to change to avoid a race condition. Similar story in WriteByte.

Reply:
I've raised a similar concern while working on MARS stream object. The consensus so far seems to be that because we explicitly state that SqlClient isn't thread-safe (I might be overstating this), we really don't need to complicate matters by synchronizing reads/writes. Though I personally think this is a good opportunity to make it properly thread-safe without mucking it up. On the other hand, if we make it unnecessarily thread-safe it will be a lot harder to prove it is still satisfactorily thread-safe (or not) once the thread safety mechanisms are removed.

Reply:
That's fair to a point - but if that's going to remain the case, I'd question whether it'd be right to support MARS at all. One of the central goals of MARS is to read multiple result sets in parallel, but if the library reading those sets isn't thread-safe then it feels like a potential source of bugs (particularly for large result sets which straddle multiple SNI packets.)

I'd suggest that at least the physical connection and MARS headers should be thread-safe, purely to allow the logical connections to be demultiplexed properly.

Reply:
Thanks @benrr101